browser/ephemeral_storage/ephemeral_storage_tab_helper.cc

```diff
@@ -0,0 +1,183 @@
+using content::BrowserContext;
+using content::NavigationHandle;
+using content::WebContents;
+
+namespace {
+
+content::SessionStorageNamespaceMap& session_storage_namespace_map() {
+  static base::NoDestructor<content::SessionStorageNamespaceMap>
+      session_storage_namespace_map;
+  return *session_storage_namespace_map.get();
+}
+
+content::SessionStorageNamespaceMap& local_storage_namespace_map() {
+  static base::NoDestructor<content::SessionStorageNamespaceMap>
+      local_storage_namespace_map;
+  return *local_storage_namespace_map.get();
+}
+
+std::string URLToStorageDomain(const GURL& url) {
+  std::string domain = net::registry_controlled_domains::GetDomainAndRegistry(
+      url, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
+
+  // GetDomainAndRegistry might return an empty string if this host is an IP
+  // address or a file URL.
+  if (domain.empty())
+    domain = url::Origin::Create(url.GetOrigin()).Serialize();
+
+  return domain;
+}
+
+}  // namespace
+namespace ephemeral_storage {
+
+EphemeralStorageTabHelper::~EphemeralStorageTabHelper() {}
+
+EphemeralStorageTabHelper::EphemeralStorageTabHelper(WebContents* web_contents)
+    : WebContentsObserver(web_contents) {}
+
+void EphemeralStorageTabHelper::ReadyToCommitNavigation(
+    NavigationHandle* navigation_handle) {
+  if (!base::FeatureList::IsEnabled(blink::features::kBraveEphemeralStorage))
+    return;
+  if (!navigation_handle->IsInMainFrame())
+    return;
+  if (navigation_handle->IsSameDocument())
+    return;
+
+  std::string domain = URLToStorageDomain(navigation_handle->GetURL());
+  std::string previous_domain =
+      URLToStorageDomain(web_contents()->GetLastCommittedURL());
+  if (domain == previous_domain)
+    return;
+
+  ClearEphemeralStorageIfNecessary();
+
+  auto* browser_context = web_contents()->GetBrowserContext();
+  auto site_instance = content::SiteInstance::CreateForURL(
+      browser_context, navigation_handle->GetURL());
+  auto* partition =
+      BrowserContext::GetStoragePartition(browser_context, site_instance.get());
+
+  std::string local_partition_id = domain + "/ephemeral-local-storage";
+  content::SessionStorageNamespaceMap::const_iterator it =
+      local_storage_namespace_map().find(local_partition_id);
+  if (it == local_storage_namespace_map().end()) {
+    auto local_storage_namespace =
+        content::CreateSessionStorageNamespace(partition, local_partition_id);
+    local_storage_namespace_map()[local_partition_id] =
+        std::move(local_storage_namespace);
+  }
+
+  std::string session_partition_id =
+      content::GetSessionStorageNamespaceId(web_contents()) +
+      "/ephemeral-session-storage";
+  it = session_storage_namespace_map().find(session_partition_id);
+  if (it == session_storage_namespace_map().end()) {
+    auto session_storage_namespace =
+        content::CreateSessionStorageNamespace(partition, session_partition_id);
+    session_storage_namespace_map()[session_partition_id] =
```

Comment:
I'm not a C++ guru but this probably should be `emplace`

Reply:
:+1: